rule-engine/rule-engine-components/src/main/java/org/thingsboard/rule/engine/action/TbAbstractCustomerActionNode.java

```diff
@@ -62,77 +69,76 @@ public void init(TbContext ctx, TbNodeConfiguration configuration) throws TbNode
 
     @Override
     public void onMsg(TbContext ctx, TbMsg msg) {
+        var entityType = msg.getOriginator().getEntityType();
+        if (!supportedEntityTypes.contains(entityType)) {
+            throw new RuntimeException(unsupportedOriginatorTypeErrorMessage(entityType));
+        }
         withCallback(processCustomerAction(ctx, msg),
                 m -> ctx.tellSuccess(msg),
-                t -> ctx.tellFailure(msg, t), ctx.getDbCallbackExecutor());
+                t -> ctx.tellFailure(msg, t), MoreExecutors.directExecutor());
     }
 
-    private ListenableFuture<Void> processCustomerAction(TbContext ctx, TbMsg msg) {
-        ListenableFuture<CustomerId> customerIdFeature = getCustomer(ctx, msg);
-        return Futures.transform(customerIdFeature, customerId -> {
-                    doProcessCustomerAction(ctx, msg, customerId);
-                    return null;
-                }, ctx.getDbCallbackExecutor()
-        );
-    }
-
-    protected abstract void doProcessCustomerAction(TbContext ctx, TbMsg msg, CustomerId customerId);
-
-    protected ListenableFuture<CustomerId> getCustomer(TbContext ctx, TbMsg msg) {
-        String customerTitle = TbNodeUtils.processPattern(this.config.getCustomerNamePattern(), msg);
-        CustomerKey key = new CustomerKey(customerTitle);
-        return ctx.getDbCallbackExecutor().executeAsync(() -> {
-            Optional<CustomerId> customerId = customerIdCache.get(key);
-            if (!customerId.isPresent()) {
-                throw new RuntimeException("No customer found with name '" + key.getCustomerTitle() + "'.");
+    protected abstract ListenableFuture<Void> processCustomerAction(TbContext ctx, TbMsg msg);
+
+    protected ListenableFuture<CustomerId> getCustomerIdFuture(TbContext ctx, TbMsg msg) {
+        var tenantId = ctx.getTenantId();
+        var customerTitle = TbNodeUtils.processPattern(this.config.getCustomerNamePattern(), msg);
+        ListenableFuture<Optional<Customer>> customerByTitleFuture = findCustomerByTitleAsync(ctx, customerTitle);
+        if (createCustomerIfNotExists()) {
+            return Futures.transform(customerByTitleFuture, customerOpt -> {
+                if (customerOpt.isPresent()) {
+                    return customerOpt.get().getId();
+                }
+                var customerCreationLockKey = new CustomerCreationLockKey(tenantId, customerTitle);
+                synchronized (customerCreationLocks.computeIfAbsent(customerCreationLockKey, k -> new Object())) {
```

Comment:
I'm not happy with blocking threads as well. However, we need to do this to make sure that the process of customer creation is atomic. BTW, we are blocking a thread from db callback executor, not a rule dispatcher thread, and this will happen only when customer wasn't found by initial call at line 86. 

This logic was approved by both @ViacheslavKlimov and @dskarzh. I don't know a better solution. If you know one, please let me know.

Reply:
synchronized will not help in cluster mode.